### PR TITLE
Improve support for Chinese languages in Languages helper

### PR DESCRIPTION
**Pull request by koke**
This fixes a couple issues:

- Apple uses `language-script` for Chinese languages, but we only do `language-region`. We already do a similar map on `update-translations.rb` to match the `.lproj` codes with GlotPress. I think this only affects the Chinese variants so far, so I added a couple exceptions, as properly parsing language tags was not a trivial task.

- While developing this, I realized that not only Apple did `language-script`, but it also appended the region. Before this PR, we'd search the full tag (`zh-Hans-ES`) and then the language (`zh`), and none of them matched the right language. Now we try the full tag and start dropping components from the end until we find a match (`zh-Hans-ES`, `zh-Hans`, `zh`, `en`)

Testing:

Try creating a new site with your device set to different languages, make sure the site has the right language.

Needs review: @jleandroperez

### CODE REVIEW

File: WordPress/Classes/Utility/Languages.swift [hunk 1]
```diff
@@ -143,17 +146,49 @@ class Languages : NSObject
...
+
+    /// Provides a sequence of language tags from the specified string, from more to less specific
+    /// For instance, "zh-Hans-HK" will yield `["zh-Hans-HK", "zh-Hans", "zh"]`
+    ///
+    private struct LanguageTagVariants: SequenceType {
+        let string: String
+
+        func generate() -> AnyGenerator<String> {
+            var components = string.componentsSeparatedByString("-")
+            return AnyGenerator {
```
Comment (jleandroperez) [inline WordPress/Classes/Utility/Languages.swift]: Love this kung fu ❤️

File: WordPress/Classes/Utility/Languages.swift [hunk 2]
```diff
@@ -68,26 +68,29 @@ class Languages : NSObject
...
-        }
-        let languageTag = deviceLanguageCode.componentsSeparatedByString("-")[0]
-        if languageTag != deviceLanguageCode,
-            let partialMatch = languageWithSlug(languageTag) {
-            return partialMatch.languageId
+        let variants = LanguageTagVariants(string: deviceLanguageCode)
+        for variant in variants {
+            if let match = self.languageWithSlug(variant) {
+                return match.languageId
+            }
```
Comment (jleandroperez) [inline WordPress/Classes/Utility/Languages.swift]: This is gonna look super cool with the upcoming `search` operator. Ahhhhh 😄

**Comment by jleandroperez** [comment]
Looking great @koke. Tested creating few blogs in Chineses (Two variants) / Spanish, and effectively, the language is reflected properly on the web.

:shipit:

File: WordPress/Classes/Utility/Languages.swift (hunk 2 above)
Comment (koke) [inline WordPress/Classes/Utility/Languages.swift]: If you're talking about `find`, it's a slightly different case: we don't want the first variant that matches the predicate, but we want to transform it and return the first non-nil.

A more functional approach would be:

```swift
return variants.lazy.flatMap({ self.languageWithSlug($0) }).first ?? 1
```

I think I tried something like that, but I don't remember if I hit some compiler errors or just found the current version more readable.

**Pull request merged**